integration_test.go

```diff
@@ -0,0 +1,279 @@
+	time.Sleep(5 * time.Second)
+	require.Error(t, intServer.Process.Signal(syscall.Signal(0)), "Expected error when signalling stopped parent process, process has not exited")
+
+}
+
+func sendIntegrationTestRequest(sleepDuration time.Duration) (*http.Response, error) {
+	reqURL := fmt.Sprintf("http://localhost:%s%s?duration=%s", intSrvPort, intSrvPath, sleepDuration)
+	timeout := int(sleepDuration.Seconds() + 2) // add two seconds to the duration of the call to get the http timeout
+
+	client := http.Client{
+		Timeout: time.Duration(timeout) * time.Second,
+	}
+	return client.Get(reqURL)
+}
+
+func startIntegrationServer(intServer *exec.Cmd, messages chan string) error {
+	outPipe, err := intServer.StdoutPipe()
+	if err != nil {
+		close(messages)
+		return err // note process not started yet so we shouldn't have to kill intserver
+	}
+	if err := intServer.Start(); err != nil {
+		close(messages)
+		return err
+	}
+	// Forward relevant messages from stdout back to the test
+	go watchStdout(outPipe, messages)
+	return nil
+}
+
+func teardownServer(pid int) {
+	proc, err := os.FindProcess(pid)
+	if err != nil {
+		fmt.Printf("Could not find process with PID %d to clean up\n", pid)
+	}
+	if err := proc.Kill(); err != nil {
+		if !strings.Contains(err.Error(), "process already finished") {
+			fmt.Printf(
+				"Could not kill process with PID %d with err %v. Manual clean up may be required\n",
+				pid, err)
+		}
+	}
+}
+
+func waitForMessage(msgs chan string, timeout time.Duration) (string, error) {
+	to := time.NewTimer(timeout)
+	select {
+	case <-to.C:
+		return "", fmt.Errorf("timeout exceeded")
+	case msg, open := <-msgs:
+		if !open {
+			return "", fmt.Errorf("unexpected close of messages channel")
+		}
+		return msg, nil
+	}
+}
+
+func waitForServer(uri string) error {
+	reqURL, err := url.ParseRequestURI(uri)
+	if err != nil {
+		return err
+	}
+	client := http.DefaultClient
+	ctx, cancel := context.WithTimeout(context.Background(), 500*time.Millisecond)
+	defer cancel()
+	for {
+		cctx, ccancel := context.WithTimeout(ctx, 10*time.Millisecond)
+		defer ccancel()
+		req := &http.Request{
+			URL: reqURL,
+		}
+		req = req.WithContext(cctx)
+		resp, err := client.Do(req)
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-cctx.Done():
+			// request took too long, go round again
+		default:
+			if err == nil {
+				if resp.StatusCode == http.StatusOK {
+					return nil
+				}
+				return fmt.Errorf("Unexpected HTTP response from server: %v", resp.StatusCode)
+			}
+		}
+	}
+}
+
+// watchStdout watches stdout from the test integration server for specific messages
+// * the PID on start up
+// * request received notification from the HTTP handler
+func watchStdout(outPipe io.ReadCloser, messages chan string) {
+	defer close(messages)
+	scanner := bufio.NewScanner(outPipe)
+	for scanner.Scan() {
+		line := scanner.Text()
+		fmt.Println(line)
+		if pid, ok := checkforpid(line); ok {
+			messages <- pid
```

Comment:
can this deadlock?

Reply:
I hadn't considered that, I could use a buffered channel but I feel like that would just delay the problem, I don't know what the best way of stopping that from happening would be

Reply:
just create a buffer as big as expected